vision/segmentation/er_ccl_clustering.go

```diff
@@ -0,0 +1,274 @@
+	validClouds := pc.PrunePointClouds(segments.PointClouds(), erCCL.MinPtsInSegment)
+	// wrap
+	objects, err := NewSegmentsFromSlice(validClouds, "")
+	if err != nil {
+		return nil, err
+	}
+	return objects.Objects, nil
+	// this seems a bit wasteful to make segments then make more segments after filtering, but rolling with it for now
+}
+
+func pcProjection(cloud pc.PointCloud, s float64, heightIsY bool) [][]node {
+	h := int(math.Ceil((cloud.MetaData().MaxX-cloud.MetaData().MinX)/s)) + 1
+	w := int(math.Ceil((cloud.MetaData().MaxZ-cloud.MetaData().MinZ)/s)) + 1
+	if !heightIsY {
+		w = int(math.Ceil((cloud.MetaData().MaxY-cloud.MetaData().MinY)/s)) + 1
+	}
+	retVal := make([][]node, h)
+	for i := range retVal {
+		retVal[i] = make([]node, w)
+		for j, curNode := range retVal[i] {
+			curNode.label = -1
+			curNode.minHeight = 0
+			curNode.maxHeight = 0
+			curNode.i = i
+			curNode.j = j
+			retVal[i][j] = curNode
+		}
+	}
+	cloud.Iterate(0, 0, func(p r3.Vector, d pc.Data) bool {
+		i := int(math.Ceil((p.X - cloud.MetaData().MinX) / s))
+		j := int(math.Ceil((p.Z - cloud.MetaData().MinZ) / s))
+		var curNode node
+		if heightIsY {
+			curNode = retVal[i][j]
+			curNode.maxHeight = math.Max(curNode.maxHeight, p.Y)
+			curNode.minHeight = math.Min(curNode.minHeight, p.Y)
+		} else {
+			j = int(math.Ceil((p.Y - cloud.MetaData().MinY) / s))
+			curNode = retVal[i][j]
+			curNode.maxHeight = math.Max(curNode.maxHeight, p.Z)
+			curNode.minHeight = math.Min(curNode.minHeight, p.Z)
+		}
+		curNode.label = i*w + j
+		retVal[i][j] = curNode
+		return true
+	})
+	return retVal
+}
+
+func labelMapUpdate(labelMap [][]node, r int, alpha, beta, s float64) bool {
+	mapChanged := false
+
+	for i, curNodeSlice := range labelMap {
+		for j, curNode := range curNodeSlice {
+			if curNode.label == -1 {
+				// skip if no points at cell
+				continue
+			}
+			minLabel := curNode.label
+			neighbors := make([]node, 0)
+			// finding neighbors + finding min label value
+			for x := 0; x < r; x++ {
+				newI := i + x
+				if newI >= len(labelMap) {
+					break
+				}
+				for y := 0; y < r; y++ {
+					newJ := j + y
+					if newJ >= len(curNodeSlice) {
+						break
+					}
+					if newI == i && newJ == j {
+						continue // might be able to remove this because original should be in neighbors list
+					}
+					neighborNode := labelMap[newI][newJ]
+					if similarEnough(curNode, neighborNode, r, alpha, beta, s) {
+						neighbors = append(neighbors, neighborNode)
+						minLabel = int(math.Min(float64(minLabel), float64(neighborNode.label)))
+					}
+				}
+			}
+			if minLabel != curNode.label {
+				mapChanged = true
+				labelMap[curNode.i][curNode.j].label = minLabel
+			}
+			for _, neighbor := range neighbors {
+				if neighbor.label != minLabel {
+					mapChanged = true
+					labelMap[neighbor.i][neighbor.j].label = minLabel
+				}
+			}
+		}
+	}
+	return mapChanged
+}
+
+func similarEnough(curNode, neighbor node, r int, alpha, beta, s float64) bool {
```

Comment:
create a comment for this function that explains the features of this similarity function